File: src/main/java/com/soapdogg/lispInterpreter/determiner/FunctionLengthDeterminer.java
package com.soapdogg.lispInterpreter.determiner;

import com.soapdogg.lispInterpreter.constants.ReservedValuesConstants;
import com.soapdogg.lispInterpreter.datamodels.AtomNode;
import com.soapdogg.lispInterpreter.datamodels.ExpressionNode;
import com.soapdogg.lispInterpreter.datamodels.Node;
import lombok.AllArgsConstructor;

@AllArgsConstructor(staticName = "newInstance")
public class FunctionLengthDeterminer {

    public int determineFunctionLength(final Node node) {
        if (node instanceof AtomNode) {
            AtomNode atomNode = (AtomNode)node;
            return atomNode.getValue().equals(ReservedValuesConstants.NIL) ? 0 : 1;
        } else {
            ExpressionNode expressionNode = (ExpressionNode)node;
            return determineFunctionLength(expressionNode.getData()) + 1;
        }
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/determiner/FunctionLengthDeterminer.kt
package com.soapdogg.lispInterpreter.determiner

import com.soapdogg.lispInterpreter.constants.ReservedValuesConstants
import com.soapdogg.lispInterpreter.datamodels.*

class FunctionLengthDeterminer {

    fun determineFunctionLength(node: NodeV2): Int {
        return if (node is AtomNode) {
            if (node.value == ReservedValuesConstants.NIL) 0 else 1
        } else {
            val expressionNode = node as ExpressionListNode
            val lastChild = expressionNode.children[expressionNode.children.size - 1]
            return if (lastChild is AtomNode && lastChild.value == ReservedValuesConstants.NIL) {
                expressionNode.children.size - 1
            } else {
                expressionNode.children.size
            }
        }
    }
}